src/main/clojure/clara/rules/engine.cljc

```diff
@@ -391,15 +391,94 @@
 
   (description [node] (str "QueryNode -- " query)))
 
+
+(defn node-rule-names
+  [child-type node]
+  (->> node
+       (tree-seq (comp seq :children) :children)
+       (keep child-type)
+       (map :name)
+       (distinct)
+       (sort)))
+
+(defn- list-of-names
+  "Returns formatted string with correctly pluralized header and
+  list of names. Returns nil if no such node is found."
+  [singular plural names]
+  (let [msg-for-unnamed (str "  An unnamed " singular ", provide names to your "
+                             plural " if you want them to be identified here.")
+        names-string (->> names
+                          (sort)
+                          (map #(if (nil? %) msg-for-unnamed %))
+                          (map #(str "  " %))
+                          (string/join "\n"))
+        names-count (count names)]
+    (let [c (count names)]
+      (cond
+        (= names-count 1) (str  "of " singular "\n" names-string "\n")
+        (> names-count 1) (str  "of " plural "\n" names-string "\n")))))
+
+(defn- throw-constraint-exception
+  "Adds a useful error message when executing a constraint node raises an exception."
+  [error node fact env]
+  (let [exp (:constraint-exp error)
+        productions (node-rule-names :production node)
+        queries (node-rule-names :query node)
+        production-section (list-of-names "rule" "rules" productions)
+        query-section (list-of-names "query" "query" queries)
+        condition (or (:condition node) ; beta node
+                      (-> node :children first :condition) ; alpha node
+                      (-> node :children first :accum-condition) ; accum node
+                      )
```

Comment:
I ended up implementing a couple of small protocols and a couple of new functions based on your comments. We can now handle multiple different conditions containing the same constraint. Here's an example output.
```
Constraint exception.
Exception raised during execution of constraint
  (> 1 rate)
when processing fact
  #msg_test.RateToUSD{:currency "CAD", :rate nil}
with no bindings

Conditions:

1. [msg_test.RateToUSD (> 1 rate)]
   rules:
     rule1-name
     rule3-name
   queries:
     query1-name

2. [:?wind <- (clara.rules.accumulators/all) :from [msg_test.RateToUSD (> 1 rate)]]
   rules:
     rule2-name
```